src/otx/v2/adapters/torch/lightning/engine.py

```diff
@@ -200,17 +233,19 @@ def train(
         update_check = self._update_config(func_args=train_args, **kwargs)
         datamodule = self.trainer_config.pop("datamodule", None)
 
-        callbacks = self._update_callbacks(callbacks=callbacks)
-        logger = self._update_logger(logger=logger, target_path=target_path)
-
         if not hasattr(self, "trainer") or update_check:
+            self._set_device(device=device)
+            mode = "train_val" if val_dataloader is not None else "train"
+            callbacks = self._update_callbacks(callbacks=callbacks, mode=mode)
+            logger = self._update_logger(logger=logger, target_path=target_path)
             self.trainer = Trainer(
                 logger=logger,
                 callbacks=callbacks,
                 **self.trainer_config,
             )
         self.config["trainer"] = self.trainer_config
-        if optimizer is not None:
+
+        if isinstance(optimizer, Optimizer):
```

Comment:
Just a question, what happens if `optimizer==None`?

Reply:
This will look for the optimizer in the config and raise an error if it's not there.